### PR TITLE
Port splitCheckRun of worker, step 2

### PR DESCRIPTION
**Pull request by xiaobogaga**
@coocood PTAL, thanks.

### CODE REVIEW

File: tikv/raftstore/worker.go [hunk 1]
```diff
@@ -244,11 +245,30 @@ func (r *splitCheckRunner) run(t task) {
...
+	ite := reader.GetIter()
+	defer reader.Close()
+	for ite.Seek(startKey); ite.Valid(); ite.Next() {
+		item := ite.Item()
+		key := item.KeyCopy(nil)
+		if exceedEndKey(key, endKey) {
+			break
+		}
+		if value, err := item.Value(); err == nil {
+			spCheckerHost.onKv(region, &splitCheckKeyEntry{key: key, valueSize: uint64(len(value)),})
```
Comment (coocood) [inline tikv/raftstore/worker.go]: Should check the returned bool value?

Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: yes. I would fix.

File: tikv/raftstore/worker.go [hunk 2]
```diff
@@ -244,11 +245,30 @@ func (r *splitCheckRunner) run(t task) {
...
 	startKey []byte, endKey []byte) ([][]byte, error) {
-	/// Todo, currently it is a place holder
-	return nil, nil
+	txn := r.engine.NewTransaction(false)
+	reader := dbreader.NewDBReader(startKey, endKey, txn, 0)
+	ite := reader.GetIter()
+	defer reader.Close()
+	for ite.Seek(startKey); ite.Valid(); ite.Next() {
+		item := ite.Item()
+		key := item.KeyCopy(nil)
```
Comment (coocood) [inline tikv/raftstore/worker.go]: copy every key during scanSplitKeys is very expensive.
Only a few keys needed to copy.
We can avoid copy here and make sure the key is copied in split checkers

Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: ok, thanks.

File: tikv/raftstore/coprocessor.go [hunk 3]
```diff
@@ -27,7 +39,7 @@ type splitChecker interface {
 
 	/// onKv is a hook called for every kv scanned during split.
 	/// Return true to abort scan early.
-	onKv(_ *observerContext, _ splitCheckKeyEntry) bool
+	onKv(_ *observerContext, _ *splitCheckKeyEntry) bool
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: We don't need to change splitCheckKeyEntry to pointer here.

File: tikv/raftstore/coprocessor.go [hunk 4]
```diff
@@ -45,31 +57,51 @@ type splitCheckerHost struct {
...
-}
-
-func (spCheckerHost *splitCheckerHost) onKv(observerCtx *observerContext, spCheKeyEntry splitCheckKeyEntry) bool {
-	// Todo: currently it is a place holder
+	return len(spCheckerHost.checkers) == 0
+}
+
+/// onKv is a hook called for every check during split.
+/// Return true means abort early.
+func (spCheckerHost *splitCheckerHost) onKv(region *metapb.Region, spCheKeyEntry *splitCheckKeyEntry) bool {
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: use struct for `splitCheckKeyEntry` can avoid one memory allocation

Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: ok, thanks

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks.

**Comment by coocood** [comment]
LGTM

**Pull request merged**